File: backend/core/attack_detector.py
"""
Attack Detector - Detects and classifies DDoS attacks
"""

from typing import Dict, Optional
from collections import deque


class AttackDetector:
    """Advanced attack detection with classification"""

    def __init__(self):
        self.anomaly_scores = deque(maxlen=50)
        self.attack_history = deque(maxlen=100)

        # Default thresholds
        self.thresholds = {
            'happy': {'max_latency': 10, 'max_packet_loss': 1},
            'alert': {'max_latency': 50, 'max_packet_loss': 5},
            'under_attack': {'max_latency': 200, 'max_packet_loss': 20},
            'stressed': {'min_latency': 200, 'min_packet_loss': 20}
        }

    def detect(self, stats: Dict) -> Dict:
        """Detect attacks and classify them"""
        if not stats['connected']:
            return {
                'state': 'disconnected',
                'attack_detected': False,
                'attack_type': None,
                'anomaly_score': 0,
                'confidence': 0
            }

        latency = stats.get('avg_latency', -1)
        packet_loss = stats.get('avg_packet_loss', 0)

        # Handle initial state (no data yet)
        if latency < 0:
            return {
                'state': 'happy',
                'attack_detected': False,
                'attack_type': None,
                'anomaly_score': 0,
                'confidence': 0
            }

        # Determine state
        state = self._determine_state(latency, packet_loss)

        # Classify attack type
        attack_type = self._classify_attack(latency, packet_loss)

        # Calculate anomaly score
        anomaly_score = self._calculate_anomaly_score(stats)

        # Calculate confidence
        confidence = self._calculate_confidence(latency, packet_loss)

        return {
            'state': state,
            'attack_detected': attack_type is not None,
            'attack_type': attack_type,
            'anomaly_score': anomaly_score,
            'confidence': confidence
        }

    def _determine_state(self, latency: float, packet_loss: float) -> str:
        """Determine current state based on thresholds"""
        if latency < self.thresholds['happy']['max_latency'] and \
           packet_loss < self.thresholds['happy']['max_packet_loss']:
            return 'happy'

        elif latency < self.thresholds['alert']['max_latency'] and \
             packet_loss < self.thresholds['alert']['max_packet_loss']:
            return 'alert'

        elif latency < self.thresholds['under_attack']['max_latency'] and \
             packet_loss < self.thresholds['under_attack']['max_packet_loss']:
            return 'under_attack'

        else:
            return 'stressed'

    def _classify_attack(self, latency: float, packet_loss: float) -> Optional[str]:
        """Classify attack type based on patterns"""
        if packet_loss > 50:
            return "ICMP Flood / Network Saturation"
        elif packet_loss > 20 and latency > 200:
            return "UDP Flood Detected"
        elif latency > 500 and packet_loss < 10:
            return "SYN Flood / Resource Exhaustion"
        elif latency > 100 and packet_loss > 10:
            return "Mixed DDoS Attack"
        elif latency > 50:
            return "Network Congestion / Slow DDoS"

        return None

    def _calculate_anomaly_score(self, stats: Dict) -> float:
        """Calculate anomaly score (0-100 scale)"""
        score = 0.0

        baseline_lat = stats.get('baseline_latency', 0)
        current_lat = stats.get('avg_latency', 0)
        baseline_loss = stats.get('baseline_packet_loss', 0)
        current_loss = stats.get('avg_packet_loss', 0)

        # Handle invalid values
        if current_lat < 0:
            return 0.0

        # Calculate based on absolute values if no baseline yet
        if baseline_lat <= 0:
            # No baseline yet, use absolute thresholds
            if current_lat > 100:
                score += 50
            elif current_lat > 50:
                score += 25
            elif current_lat > 20:
                score += 10
        else:
            # Compare to baseline
            lat_deviation = (current_lat - baseline_lat) / baseline_lat
            if lat_deviation > 0:  # Only count increases
                score += min(lat_deviation * 30, 50)

        # Packet loss scoring
        if current_loss > 10:
            score += 50
        elif current_loss > 5:
            score += 25
        elif current_loss > 1:
            score += 10

        self.anomaly_scores.append(score)
        return min(score, 100.0)

    def _calculate_confidence(self, latency: float, packet_loss: float) -> float:
        """Calculate detection confidence"""
        confidence = 0.0

        if latency > 200:
            confidence += 40
        elif latency > 100:
            confidence += 20
        elif latency > 50:
            confidence += 10

        if packet_loss > 20:
            confidence += 60
        elif packet_loss > 10:
            confidence += 30
        elif packet_loss > 5:
            confidence += 15

        return min(confidence, 100.0)

File: ESP32_DDoS_Gotchi/attack_detector.cpp
/*
 * Attack Detector Implementation
 * Ported from backend/core/attack_detector.py
 */

#include "attack_detector.h"

// ============================================================================
// Constructor
// ============================================================================

AttackDetector::AttackDetector() {
    baselineLatency = 0.0;
    baselinePacketLoss = 0.0;
    baselinePacketRate = 0.0;
    baselineEstablished = false;
}

// ============================================================================
// Initialization
// ============================================================================

void AttackDetector::begin() {
    Serial.println("[DETECTOR] Initializing attack detector...");
    Serial.println("[DETECTOR] Learning baseline (30 samples)...");
}

// ============================================================================
// Main Analysis Function
// ============================================================================

AttackResult AttackDetector::analyze(NetworkStats& stats) {
    AttackResult result;

    // Update baseline during learning phase
    if (!baselineEstablished) {
        updateBaseline(stats);
    }

    // Detect attack state using thresholds
    result.state = detectStateByThresholds(stats);

    // Classify specific attack type
    result.attackType = classifyAttackType(stats);

    // Calculate confidence score
    result.confidence = calculateConfidence(stats, result.state);

    // Calculate anomaly score
    result.anomalyScore = calculateAnomalyScore(stats);

    // Set attack flag
    result.isUnderAttack = (result.state == STATE_UNDER_ATTACK ||
                           result.state == STATE_STRESSED);

    return result;
}

// ============================================================================
// Threshold-Based State Detection
// (Ported from attack_detector.py lines 54-83)
// ============================================================================

AttackState AttackDetector::detectStateByThresholds(NetworkStats& stats) {
    // STRESSED: Severe degradation
    if (stats.latency > LATENCY_STRESSED || stats.packetLoss > PACKET_LOSS_STRESSED) {
        return STATE_STRESSED;
    }

    // UNDER_ATTACK: Significant degradation
    if (stats.latency > LATENCY_ATTACK || stats.packetLoss > PACKET_LOSS_ATTACK) {
        return STATE_UNDER_ATTACK;
    }

    // ALERT: Warning threshold
    if (stats.latency > LATENCY_WARNING || stats.packetLoss > PACKET_LOSS_WARNING) {
        return STATE_ALERT;
    }

    // HAPPY: Normal operation
    return STATE_HAPPY;
}

// ============================================================================
// Attack Type Classification
// (Ported from attack_detector.py lines 85-98)
// ============================================================================

AttackType AttackDetector::classifyAttackType(NetworkStats& stats) {
    // ICMP Flood: Very high packet loss
    if (stats.packetLoss > PACKET_LOSS_STRESSED) {
        return ATTACK_ICMP_FLOOD;
    }

    // UDP Flood: High packet loss + high latency
    if (stats.packetLoss > PACKET_LOSS_ATTACK && stats.latency > LATENCY_ATTACK) {
        return ATTACK_UDP_FLOOD;
    }

    // SYN Flood: Very high latency but lower packet loss
    if (stats.latency > LATENCY_STRESSED && stats.packetLoss < PACKET_LOSS_WARNING) {
        return ATTACK_SYN_FLOOD;
    }

    // Mixed DDoS: Moderate latency + moderate packet loss
    if (stats.latency > 100 && stats.packetLoss > PACKET_LOSS_WARNING) {
        return ATTACK_MIXED;
    }

    // Volumetric: High packet rate (above baseline)
    if (baselineEstablished &&
        stats.packetRate > (baselinePacketRate * ANOMALY_MULTIPLIER)) {
        return ATTACK_VOLUMETRIC;
    }

    // Slow DDoS: Elevated latency
    if (stats.latency > LATENCY_WARNING) {
        return ATTACK_SLOW;
    }

    return ATTACK_NONE;
}

// ============================================================================
// Confidence Calculation
// (Ported from attack_detector.py lines 139-157)
// ============================================================================

uint8_t AttackDetector::calculateConfidence(NetworkStats& stats, AttackState state) {
    uint8_t confidence = 0;

    // Latency-based confidence
    if (stats.latency > LATENCY_STRESSED) {
        confidence += 60;
    } else if (stats.latency > LATENCY_ATTACK) {
        confidence += 40;
    } else if (stats.latency > LATENCY_WARNING) {
        confidence += 20;
    } else if (stats.latency < LATENCY_HAPPY) {
        confidence += 10;
    }

    // Packet loss-based confidence
    if (stats.packetLoss > PACKET_LOSS_STRESSED) {
        confidence += 60;
    } else if (stats.packetLoss > PACKET_LOSS_ATTACK) {
        confidence += 40;
    } else if (stats.packetLoss > PACKET_LOSS_WARNING) {
        confidence += 25;
    } else if (stats.packetLoss < PACKET_LOSS_HAPPY) {
        confidence += 15;
    }

    // Anomaly-based confidence boost
    if (isAnomalous(stats)) {
        confidence += 20;
    }

    // Cap at 100%
    if (confidence > 100) {
        confidence = 100;
    }

    return confidence;
}

// ============================================================================
// Anomaly Score Calculation
// (Ported from attack_detector.py lines 100-137)
// ============================================================================

float AttackDetector::calculateAnomalyScore(NetworkStats& stats) {
    if (!baselineEstablished) {
        return 0.0;
    }

    float score = 0.0;

    // Latency deviation
    if (baselineLatency > 0) {
        float latencyDeviation = abs(stats.latency - baselineLatency) / baselineLatency;
        score += latencyDeviation * 40.0; // Weight: 40%
    }

    // Packet loss deviation
    if (baselinePacketLoss > 0) {
        float lossDeviation = abs(stats.packetLoss - baselinePacketLoss) / baselinePacketLoss;
        score += lossDeviation * 30.0; // Weight: 30%
    } else if (stats.packetLoss > 0) {
        score += stats.packetLoss; // Direct contribution if baseline is zero
    }

    // Packet rate deviation
    if (baselinePacketRate > 0) {
        float rateDeviation = abs(stats.packetRate - baselinePacketRate) / baselinePacketRate;
        score += rateDeviation * 30.0; // Weight: 30%
    }

    // Cap at 100
    if (score > 100.0) {
        score = 100.0;
    }

    return score;
}

// ============================================================================
// Baseline Update (Learning Phase)
// ============================================================================

void AttackDetector::updateBaseline(NetworkStats& stats) {
    // Only update if we have enough history
    if (stats.historyCount < BASELINE_LEARNING_SAMPLES) {
        return;
    }

    // Calculate average from history
    float sumLatency = 0.0;
    float sumLoss = 0.0;
    float sumRate = 0.0;

    for (int i = 0; i < stats.historyCount; i++) {
        sumLatency += stats.latencyHistory[i];
        sumLoss += stats.packetLossHistory[i];
        sumRate += stats.packetRateHistory[i];
    }

    baselineLatency = sumLatency / stats.historyCount;
    baselinePacketLoss = sumLoss / stats.historyCount;
    baselinePacketRate = sumRate / stats.historyCount;

    baselineEstablished = true;

    Serial.println("\n[DETECTOR] ✓ Baseline established:");
    Serial.print("           Latency: ");
    Serial.print(baselineLatency, 2);
    Serial.println(" ms");
    Serial.print("           Packet Loss: ");
    Serial.print(baselinePacketLoss, 2);
    Serial.println("%");
    Serial.print("           Packet Rate: ");
    Serial.print(baselinePacketRate, 1);
    Serial.println(" pps\n");
}

// ============================================================================
// Check if Current Stats are Anomalous
// ============================================================================

bool AttackDetector::isAnomalous(NetworkStats& stats) {
    if (!baselineEstablished) {
        return false;
    }

    // Check if any metric exceeds anomaly threshold
    bool latencyAnomaly = (stats.latency > baselineLatency * ANOMALY_MULTIPLIER);
    bool lossAnomaly = (stats.packetLoss > baselinePacketLoss * ANOMALY_MULTIPLIER);
    bool rateAnomaly = (stats.packetRate > baselinePacketRate * ANOMALY_MULTIPLIER);

    return (latencyAnomaly || lossAnomaly || rateAnomaly);
}
